utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,195 @@
+     * @param searchInterval for searching. Searcher stops to send requests when the next `from` time is before than search start time minus [searchInterval]
+     */
+    @JvmOverloads
+    fun findLastOrNull(
+        book: String = defaultBook,
+        sessionGroup: String = defaultSessionGroup,
+        messageStreams: Set<MessageStream> = defaultMessageStreams,
+        searchInterval: Duration,
+        filter: Function<MessageSearchResponse, Boolean>
+    ): MessageSearchResponse? {
+        require(book.isNotBlank()) {
+            "'book' can't be blank"
+        }
+        require(sessionGroup.isNotBlank()) {
+            "'sessionGroup' can't be blank"
+        }
+        require(messageStreams.isNotEmpty()) {
+            "'messageStreams' can't be empty"
+        }
+        messageStreams.forEach {
+            require(it.name.isNotBlank()) {
+                "'name' in ${it.toJson()} message stream can't be blank"
+            }
+            require(it.direction != UNRECOGNIZED) {
+                "'direction' in ${it.toJson()} message stream can't be $UNRECOGNIZED"
+            }
+        }
+
+        require(!searchInterval.isZero) {
+            "'searchInterval' can't be zero"
+        }
+        require(!searchInterval.isNegative) {
+            "'searchInterval' can't be negative"
+        }
+
+        val now = Instant.now()
+        val end = now.minus(searchInterval)
+        var from: Instant
+        var to = now
+
+        val requestBuilder = createRequestBuilder(
+            book = book,
+            sessionGroup = sessionGroup,
+            messageStreams = messageStreams
+        )
+        K_LOGGER.info { "Search the last or null, base request: ${requestBuilder.toJson()}, start time: $now, end time: $end" }
+
+        return withCancellation {
+            generateSequence {
+                if (to.isBefore(end)) {
+                    return@generateSequence null
+                }
+
+                from = to
+                to = from.minus(searchStep)
+
+                service.searchMessageGroups(
+                    requestBuilder.apply {
+                        startTimestamp = from.toTimestamp()
+                        endTimestamp = to.toTimestamp()
+                    }.build().also {
+                        K_LOGGER.debug { "Request from: $from, to: $to" }
+                    }
+                )
+            }.flatMap(Iterator<MessageSearchResponse>::asSequence)
+                .onEach { K_LOGGER.trace { "Message: ${it.toJson()}" } }
+                .firstOrNull(filter::apply)
+        }
+    }
+
+    @JvmOverloads
+    fun with(
+        requestInterval: Duration = DEFAULT_SEARCH_STEP,
+        defaultBook: String = "",
+        defaultSessionGroup: String = "",
+        defaultMessageStreams: Set<MessageStream> = emptySet(),
+    ) = MessageSearcher(service, requestInterval, defaultBook, defaultSessionGroup, defaultMessageStreams)
```

Comment:
If all parameters have a default value we will have an overload with zero parameters. Do we need such overload? Maybe this method can be made private - we have all other methods to set those values. I don't think it really matters whether we create one copy of the object or several in the process here

Reply:
`Do we need such overload?` this method can be used for complex change or complex clear MessageSearcher. The last one doesn't broke logic but maybe useful in specific cases

Reply:
Why don't just provide a method `withDefaults` to create a copy with default values for all parameters?
I do not doubt that the method with all the parameters will be helpful. But when you call `val copy = searcher.with()` what do you expect? I, honestly, have no idea what to expect as a result when I see such a method call. And definitely, I don't expect that all the parameters will be reset to their defaults.